numpy/core/src/multiarray/compiled_base.c

```diff
@@ -57,66 +57,101 @@ check_array_monotonic(const double *a, npy_int lena)
+    PyArray_Descr *dtype = PyArray_DescrFromType(type);
+    return (PyArrayObject *)PyArray_CheckFromAny(object, dtype, 1, 1,
+                                                 flags | extra_flags, NULL);
 }
 
 /*
  * arr_bincount is registered as bincount.
  *
- * bincount accepts one, two or three arguments. The first is an array of
- * non-negative integers The second, if present, is an array of weights,
- * which must be promotable to double. Call these arguments list and
- * weight. Both must be one-dimensional with len(weight) == len(list). If
- * weight is not present then bincount(list)[i] is the number of occurrences
- * of i in list.  If weight is present then bincount(self,list, weight)[i]
- * is the sum of all weight[j] where list [j] == i.  Self is not used.
+ * bincount accepts one, two, three or four arguments. The first is an
+ * array of non-negative integers.  The second, if present, is an array
+ * of weights, which must be promotable to double.  Call these arguments
+ * list and weights.  Both must be one-dimensional with the same length.
+ * If weights is not present then bincount(list)[i] is the number of
+ * occurrences of i in list.  If weights is present then
+ * bincount(self,list, weight)[i] is the sum of all weight[j] where
+ * list[j] == i.  Self is not used.
  * The third argument, if present, is a minimum length desired for the
  * output array.
+ * The fourth, if present, is an array of the right type and size on
+ * which to add the counts or weights sums.
  */
 NPY_NO_EXPORT PyObject *
 arr_bincount(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwds)
 {
-    PyObject *list = NULL, *weight = Py_None, *mlength = Py_None;
-    PyArrayObject *lst = NULL, *ans = NULL, *wts = NULL;
-    npy_intp *numbers, *ians, len, mx, mn, ans_size, minlength;
-    npy_intp i;
-    double *weights , *dans;
-    static char *kwlist[] = {"list", "weights", "minlength", NULL};
-
-    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|OO:bincount",
-                kwlist, &list, &weight, &mlength)) {
-            goto fail;
+    PyObject *list_obj = NULL;
+    PyObject *weights_obj = Py_None;
+    PyObject *minlength_obj = Py_None;
+    PyObject *out_obj = Py_None;
+    PyArrayObject *list_arr = NULL;
+    PyArrayObject *weights_arr = NULL;
+    npy_intp minlength = 0;
+    PyArrayObject *out_arr = NULL;
+    npy_intp len, list_stride, list_min, list_max;
+    npy_intp weights_stride;
+    npy_intp out_len, out_stride;
+    const char *list_ptr;
+    const char *weights_ptr;
+    char *out_ptr;
+
+    static char *kwlist[] = {"list", "weights", "minlength", "out", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|OOO:bincount", kwlist,
+                         &list_obj, &weights_obj, &minlength_obj, &out_obj)) {
+        goto fail;
     }
 
-    lst = (PyArrayObject *)PyArray_ContiguousFromAny(list, NPY_INTP, 1, 1);
-    if (lst == NULL) {
+    list_arr = _array_from_object(list_obj, NPY_INTP, 0);
```

Comment:
Hmm... Perhaps we could mix in here some of the ideas in #7464, which I apparently abandoned for no good reason. Will give it some thought.